drivers/nec/display/display.cr

```diff
@@ -0,0 +1,397 @@
+      audio_input
+    end
+  end
+
+  private def parse_response(data : String)
+    # 14..15 == type (we don't care)
+    max = data[16..19].to_i(16)
+    value = data[20..23].to_i(16)
+
+    case Command.from_value(data[10..13].to_i(16))
+    when .video_input?
+      input = Inputs.from_value(value)
+      self[:input] = input
+      target_input = @target_input ||= input
+      if target_input && self[:input] != self[:target_input]
+        switch_to(target_input)
+      end
+    when .audio_input?
+      self[:audio] = Audio.from_value(value)
+      target_audio = @target_audio
+      if target_audio && self[:audio] != self[:target_audio]
+        switch_audio(target_audio)
+      end
+    when .volume_status?
+      self[:volume_max] = max
+      unless self[:audio_mute]
+        self[:volume] = value
+      end
+    when .brightness_status?
+      self[:brightness_max] = max
+      self[:brightness] = value
+    when .contrast_status?
+      self[:contrast_max] = max
+      self[:contrast] = value
+    when .mute_status?
+      self[:audio_mute] = value == 1
+      if(value == 1)
+        self[:volume] = 0
+      else
+        volume_status(60) # high priority
+      end
+    when .power_on_delay?
+      if value > 0
+        self[:warming] = true
+        schedule.in(value.seconds) do # Prevent any commands being sent until the power on delay is complete
+          power_on_delay.as(PlaceOS::Driver::Transport)
+        end
+      else
+        schedule.in(3.seconds) do # Reactive the interface once the display is online
+          self[:warming] = false # allow access to the display
+        end
+      end
+    when .auto_setup?
+      # auto_setup
+      # nothing needed to do here (we are delaying the next command by 4 seconds)
+    else
+      logger.info { "-- NEC LCD, unknown response: #{data[10..13]}" }
+      logger.info { "-- NEC LCD, full response was: #{data}" }
+    end
+  end
+
+  # Types of messages sent to and from the LCD
+  enum MsgType
+    Command             = 0x41 # 'A'
+    Command_reply       = 0x42 # 'B'
+    Get_parameter       = 0x43 # 'C'
+    Get_parameter_reply = 0x44 # 'D'
+    Set_parameter       = 0x45 # 'E'
+    Set_parameter_reply = 0x46 # 'F'
+  end
+
+  enum Command
+    Video_input       = 0x0060
+    Audio_input       = 0x022E
+    Volume_status     = 0x0062
+    Mute_status       = 0x008D
+    Power_on_delay    = 0x02D8
+    Contrast_status   = 0x0012
+    Brightness_status = 0x0010
+    Auto_setup        = 0x001E
+    Power_query       = 0x01D6
+
+    def to_s : String
+      Nec::Display::All.format_value(self.value)
+    end
+  end
+
+  def self.format_value(value : Int, length : Int = 4) : String
+    value.to_s(16).upcase.rjust(length, '0')
+  end
+
+  {% for name in Command.constants.map(&.downcase) %}
+  @[Security(Level::Administrator)]
+    def {{name.id}}(priority : Int32 = 0)
+      do_send(MsgType::Get_parameter, Command.parse({{name}}), priority: priority, name: {{name.id}})
```

Comment:
You shouldn't need any runtime parsing here. This can just be `Command::{{name.id}}` if you move the name mapping to the method naming only.

If you end up changing those enum names to camelcase too, I'd recommend [`String#underscore`](https://crystal-lang.org/api/0.35.1/String.html#underscore(options:Unicode::CaseOptions=:none):String-instance-method) for helping with this.

Reply:
Okay I've made this commit which might still not be the best way but I couldn't figure out a way to only call `underscore` once and I can't use do `name: {{name.id}}` without getting an error in the driver tester

https://github.com/PlaceOS/drivers/pull/50/commits/c78a02894f37b391da953e4799600c6159e808e5

Reply:
The `name` param is used by the underlying Queue to make sure only one instance of the associated command is stacked up at any point. It just needs to be something that can be mapped into a `String`.

At the moment when that macro expands you will be left with `name: video_input` etc. `video_input` is not a valid identifier within you current scope so the compiler will complain about that. Tack on a [`.symbolize`](https://crystal-lang.org/api/0.35.1/Crystal/Macros/ASTNode.html#symbolize:SymbolLiteral-instance-method) or [`.stringify`](https://crystal-lang.org/api/0.35.1/Crystal/Macros/ASTNode.html#stringify:StringLiteral-instance-method) and it will expand to what you're likely after there.

Also, don't stress a great deal about perf / minimising operations within macros. While this is definitely a good goal, it's a cost which is only incurred once as the driver compiles. It's only the expanded code that contributes to run-time cost.

Reply:
nice, thanks for that explanation @KimBurgess 

The reason I did 
`      do_send(MsgType::GetParameter, Command::{{name.id}}, priority: priority, name: {{name.id.underscore}})`

and not
`      do_send(MsgType::GetParameter, Command::{{name.id}}, priority: priority, name: {{name.id.underscore.stringify}})`

is because the later actually gives an error in the driver runner even though this evaluates fine in crystal play as well as other alternatives I tried https://play.crystal-lang.org/#/r/9r2l

This is the error that I get:
```In drivers/nec/display/display.cr:325:11

 325 | value.to_s(16, true).rjust(length, '0')
             ^---
Warning: Deprecated Int32#to_s. Use `#to_s(base : Int, *, upcase : Bool = false)` instead

A total of 1 warnings were found.
In lib/placeos-driver/src/placeos-driver.cr:390:1

 390 | macro finished
       ^
Error: expanding macro


There was a problem expanding macro 'finished'

Called macro defined in lib/placeos-driver/src/placeos-driver.cr:390:1

 390 | macro finished

Which expanded to:

    1 |   exec_process_manager = false
    2 | 
    3 |   # Command line options
    4 |   OptionParser.parse(ARGV.dup) do |parser|
    5 |     parser.banner = "Usage: #{PROGRAM_NAME} [arguments]"
    6 | 
    7 |     parser.on("-m", "--metadata", "output driver metadata") do
    8 |       puts Nec::Display::All::KlassExecutor.metadata
    9 |       exit 0
   10 |     end
   11 | 
   12 |     parser.on("-d", "--defaults", "output driver defaults") do
   13 |       puts PlaceOS::Driver::Utilities::Discovery.defaults
   14 |       exit 0
   15 |     end
   16 | 
   17 |     parser.on("-p", "--process", "starts the process manager (expects to have been launched by PlaceOS core)") do
   18 |       exec_process_manager = true
   19 |     end
   20 | 
   21 |     parser.on("-h", "--help", "show this help") do
   22 |       puts parser
   23 |       exit 0
   24 |     end
   25 |   end
   26 | 
   27 |   # Launch the process manager by default, this can be overriten for testing
   28 |   if exec_process_manager
 > 29 |     process = PlaceOS::Driver::ProcessManager.new
   30 | 
   31 |     # Detect ctr-c to shutdown gracefully
   32 |     Signal::INT.trap do |signal|
   33 |       puts " > terminating gracefully"
   34 |       spawn(same_thread: true) { process.terminate }
   35 |       signal.ignore
   36 |     end
   37 | 
   38 |     process.terminated.receive?
   39 |   end
Error: instantiating 'PlaceOS::Driver::ProcessManager.class#new()'


In lib/placeos-driver/src/placeos-driver/process_manager.cr:7:3

 7 | def initialize(@logger_io = ::PlaceOS::Driver.logger_io, @input = STDIN, output = STDERR)
     ^---------
Error: instantiating 'initialize(IO+, IO::FileDescriptor, IO::FileDescriptor)'


In lib/placeos-driver/src/placeos-driver/process_manager.cr:17:43

 17 | @protocol.register :start { |request| start(request) }
                                            ^----
Error: instantiating 'start(PlaceOS::Driver::Protocol::Request)'


In lib/placeos-driver/src/placeos-driver/process_manager.cr:33:3

 33 | def start(request : Protocol::Request, driver_model = nil)
      ^----
Error: instantiating 'start(PlaceOS::Driver::Protocol::Request, Nil)'


In lib/placeos-driver/src/placeos-driver/process_manager.cr:38:28

 38 | driver = DriverManager.new module_id, model, @logger_io, @subscriptions
                             ^--
Error: instantiating 'PlaceOS::Driver::DriverManager.class#new(String, PlaceOS::Driver::DriverModel, IO+, PlaceOS::Driver::Subscriptions)'


In lib/placeos-driver/src/placeos-driver/driver_manager.cr:21:22

 21 | received(data, task)
      ^-------
Error: instantiating 'received(Slice(UInt8), (PlaceOS::Driver::Task | Nil))'


In lib/placeos-driver/src/placeos-driver/driver_manager.cr:207:14

 207 | driver.received(data, task)
              ^-------
Error: instantiating 'Nec::Display::All#received(Slice(UInt8), (PlaceOS::Driver::Task | Nil))'


In drivers/nec/display/display.cr:217:9

 217 | parse_response(ascii_string)
       ^-------------
Error: instantiating 'parse_response(String)'


In drivers/nec/display/display.cr:282:11

 282 | power_on_delay.as(PlaceOS::Driver::Transport)
       ^
Error: can't cast PlaceOS::Driver::Task to PlaceOS::Driver::Transport```

Reply:
nevermind, ignore this just some bad coding on my end 😞 
https://github.com/PlaceOS/drivers/pull/50/commits/6e81e0cd100effb0f474626ad1a541c53cc1e8a0